### PR TITLE
Fix case where empty HTTP response caused panic in reconcile loop

### LINKED ISSUE #88: Panic in reconcile loop
## Describe the bug
The messaging topology operator Pod is in crash loop backoff. 

## To Reproduce
I haven't reproduced the problem in KinD. The setup I have is:

- Two RabbitMQ clusters in two different namespaces
- Schema sync plugin enabled. One cluster synchronises from the other
- User `schema-sync` CR created in both namespaces
- User object `schema-sync` has a deletion timestamp in its metadata
- Topology Operator is getting a 401 error when it contacts RabbitMQ API

**Expected behavior**
The topology operator handles the response from the API and doesn't crash.

## Screenshots

Log fragment of the panic

```
2021-03-29T11:07:13.337Z        INFO    controller-runtime.manager.controller.user      Deleting        {"reconciler group": "rabbitmq.com", "reconciler kind": "User", "name": "schema-sync", "namespace": "aitor-secondary"}
2021-03-29T11:07:13.337Z        INFO    controller-runtime.manager.controller.user      Successfully retrieved credentials      {"reconciler group": "rabbitmq.com", "reconciler kind": "User", "name": "schema-sync", "namespace": "aitor-secondary", "user": "schema-sync", "secretCredentials": "&LocalObjectReference{Name:schema-sync-user-credentials,}"}
2021-03-29T11:07:13.337Z        INFO    controller-runtime.manager.controller.queue     Start reconciling       {"reconciler group": "rabbitmq.com", "reconciler kind": "Queue", "name": "invoices", "namespace": "aitor-primary", "spec": "{\"name\":\"invoices\",\"vhost\":\"/\",\"type\":\"quorum\",\"durable\":true,\"rabbitmqClusterReference\":{\"name\":\"aitor-primary\"}}"}
2021-03-29T11:07:13.337Z        INFO    controller-runtime.manager.controller.exchange  Start reconciling       {"reconciler group": "rabbitmq.com", "reconciler kind": "Exchange", "name": "x-test", "namespace": "aitor-primary", "spec": "{\"name\":\"x-test\",\"vhost\":\"/\",\"type\":\"direct\",\"durable\":true,\"rabbitmqClusterReference\":{\"name\":\"aitor-primary\"}}"}
E0329 11:07:13.341975       1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 425 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x157ae40, 0x22b2540)
        /go/pkg/mod/k8s.io/apimachinery@v0.20.5/pkg/util/runtime/runtime.go:74 +0x95
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /go/pkg/mod/k8s.io/apimachinery@v0.20.5/pkg/util/runtime/runtime.go:48 +0x86
panic(0x157ae40, 0x22b2540)
        /usr/local/go/src/runtime/panic.go:965 +0x1b9
github.com/rabbitmq/messaging-topology-operator/controllers.validateResponseForDeletion(0x0, 0x1942de0, 0xc0006e2d40, 0x0, 0x1942de0)
        /workspace/controllers/utils.go:131 +0x29
github.com/rabbitmq/messaging-topology-operator/controllers.(*UserReconciler).deleteUser(0xc0002ab5c0, 0x1973118, 0xc000795830, 0xc000079140, 0xc00045d980, 0x0, 0xc000387a00)
        /workspace/controllers/user_controller.go:320 +0x327
github.com/rabbitmq/messaging-topology-operator/controllers.(*UserReconciler).Reconcile(0xc0002ab5c0, 0x1973118, 0xc000795830, 0xc000387a00, 0xf, 0xc0003879e0, 0xb, 0xc000795830, 0xc000030000, 0x15f8080, ...)
        /workspace/controllers/user_controller.go:77 +0xed5
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc000422140, 0x1973070, 0xc00037e340, 0x15c4b00, 0xc0007393c0)
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:298 +0x30d
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000422140, 0x1973070, 0xc00037e340, 0x200c000132c00)
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:253 +0x205
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.2(0x1973070, 0xc00037e340)
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:216 +0x4a
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1()
        /go/pkg/mod/k8s.io/apimachinery@v0.20.5/pkg/util/wait/wait.go:185 +0x37
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc000028f50)
        /go/pkg/mod/k8s.io/apimachinery@v0.20.5/pkg/util/wait/wait.go:155 +0x5f
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc0005e5f50, 0x19446e0, 0xc0007957a0, 0xc00037e301, 0xc0005b2ae0)
        /go/pkg/mod/k8s.io/apimachinery@v0.20.5/pkg/util/wait/wait.go:156 +0x9b
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000028f50, 0x3b9aca00, 0x0, 0xc000172f01, 0xc0005b2ae0)
        /go/pkg/mod/k8s.io/apimachinery@v0.20.5/pkg/util/wait/wait.go:133 +0x98
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext(0x1973070, 0xc00037e340, 0xc0006e28c0, 0x3b9aca00, 0x0, 0xc000313801)
        /go/pkg/mod/k8s.io/apimachinery@v0.20.5/pkg/util/wait/wait.go:185 +0xa6
k8s.io/apimachinery/pkg/util/wait.UntilWithContext(0x1973070, 0xc00037e340, 0xc0006e28c0, 0x3b9aca00)
        /go/pkg/mod/k8s.io/apimachinery@v0.20.5/pkg/util/wait/wait.go:99 +0x57
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:213 +0x40d
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x1469389]

goroutine 425 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /go/pkg/mod/k8s.io/apimachinery@v0.20.5/pkg/util/runtime/runtime.go:55 +0x109
panic(0x157ae40, 0x22b2540)
        /usr/local/go/src/runtime/panic.go:965 +0x1b9
github.com/rabbitmq/messaging-topology-operator/controllers.validateResponseForDeletion(0x0, 0x1942de0, 0xc0006e2d40, 0x0, 0x1942de0)
        /workspace/controllers/utils.go:131 +0x29
github.com/rabbitmq/messaging-topology-operator/controllers.(*UserReconciler).deleteUser(0xc0002ab5c0, 0x1973118, 0xc000795830, 0xc000079140, 0xc00045d980, 0x0, 0xc000387a00)
        /workspace/controllers/user_controller.go:320 +0x327
github.com/rabbitmq/messaging-topology-operator/controllers.(*UserReconciler).Reconcile(0xc0002ab5c0, 0x1973118, 0xc000795830, 0xc000387a00, 0xf, 0xc0003879e0, 0xb, 0xc000795830, 0xc000030000, 0x15f8080, ...)
        /workspace/controllers/user_controller.go:77 +0xed5
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc000422140, 0x1973070, 0xc00037e340, 0x15c4b00, 0xc0007393c0)
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:298 +0x30d
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000422140, 0x1973070, 0xc00037e340, 0x200c000132c00)
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:253 +0x205
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.2(0x1973070, 0xc00037e340)
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:216 +0x4a
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1()
        /go/pkg/mod/k8s.io/apimachinery@v0.20.5/pkg/util/wait/wait.go:185 +0x37
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc000028f50)
        /go/pkg/mod/k8s.io/apimachinery@v0.20.5/pkg/util/wait/wait.go:155 +0x5f
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc0005e5f50, 0x19446e0, 0xc0007957a0, 0xc00037e301, 0xc0005b2ae0)
        /go/pkg/mod/k8s.io/apimachinery@v0.20.5/pkg/util/wait/wait.go:156 +0x9b
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000028f50, 0x3b9aca00, 0x0, 0xc000172f01, 0xc0005b2ae0)
        /go/pkg/mod/k8s.io/apimachinery@v0.20.5/pkg/util/wait/wait.go:133 +0x98
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext(0x1973070, 0xc00037e340, 0xc0006e28c0, 0x3b9aca00, 0x0, 0xc000313801)
        /go/pkg/mod/k8s.io/apimachinery@v0.20.5/pkg/util/wait/wait.go:185 +0xa6
k8s.io/apimachinery/pkg/util/wait.UntilWithContext(0x1973070, 0xc00037e340, 0xc0006e28c0, 0x3b9aca00)
        /go/pkg/mod/k8s.io/apimachinery@v0.20.5/pkg/util/wait/wait.go:99 +0x57
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:213 +0x40d
```

## Version and environment information
- Messaging Topology Operator: 0.4.1
- RabbitMQ: 3.8.12
- RabbitMQ Cluster Operator: 1.5.0
- Kubernetes: 1.18
- Cloud provider or hardware configuration: GKE

### PR DESCRIPTION
**Pull request by coro**
This closes #88 

**Note to reviewers:** remember to look at the commits in this PR and consider if they can be squashed

## Summary Of Changes

Both `validateResponse` and `validateResponseForDeletion` can panic if they are asked to validate a nil-reponse object. Rabbithole returns this in some cases, where there is a Go error or a 401 from the API.

For both of these, the function now returns an error if both `err` and the response is nil.

## Additional Context

This also adds in integration testing for one of the controllers, and the infrastructure to allow this to be done on the other controllers in the future.

### CODE REVIEW

File: controllers/user_controller_test.go [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+				},
+			}
+		})
+
+		Context("Creating a user", func() {
+			When("the RabbitMQ Client returns a HTTP error response", func() {
+				BeforeEach(func() {
+					userName = "test-user-http-error"
+					fakeRabbitMQClient.PutUserReturns(&http.Response{
+						Status:     "418 I'm a teapot",
```
Comment (ChunyiLyu) [inline controllers/user_controller_test.go:44]: great touch!

Comment (coro) [reply controllers/user_controller_test.go:44]: 🫖

File: controllers/utils.go [hunk 2]
```diff
@@ -36,16 +39,8 @@ func validateResponse(res *http.Response, err error) error {
...
-	if err != nil {
-		return err
-	}
-	if res.StatusCode >= http.StatusMultipleChoices {
-		body, _ := ioutil.ReadAll(res.Body)
-		res.Body.Close()
-		return fmt.Errorf("request failed with status code %d and body %q", res.StatusCode, body)
-	}
-	return nil
+	return validateResponse(res, err)
```
Comment (ChunyiLyu) [inline controllers/utils.go:45]: Smart refactor!

**Review by ChunyiLyu — APPROVED** [review]
Awesome 🚢

File: controllers/controllers_suite_test.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/envtest/printer"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+)
+
+func TestControllers(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecsWithDefaultAndCustomReporters(t,
+		"Controller Suite",
+		[]Reporter{printer.NewlineReporter{}})
```
Comment (ChunyiLyu) [inline controllers/controllers_suite_test.go:38]: May I ask why do we need a custom reporter here?

Comment (coro) [reply controllers/controllers_suite_test.go:38]: Taken from the cluster operator controller suite, but I'm not sure why it's used there either. It's also getting removed in Ginkgo 2.0, so i'll change it to use the default

Comment (ChunyiLyu) [reply controllers/controllers_suite_test.go]: sounds great!

**Review by Zerpet — APPROVED** [review]
Hot! 🔥

**Pull request merged**